### ISSUE TITLE
leak in coda-definition-parse.c:parse_entry

### ISSUE THREAD

**Issue by ruimaranhao** [stcorp/coda#80]
Use of uninitialized variable found, whcih can be reproduced using the test case below:

```
==45769==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x557820a68700 in normal_scanPoundName third_party/expat/lib/xmltok_impl.c:953:13
    #1 0x557820a5eb3d in normal_prologTok third_party/expat/lib/xmltok_impl.c:1132:12
    #2 0x557820a813e6 in initScan third_party/expat/lib/xmltok.c:1640:10
    #3 0x557820a5e266 in initScanPrologNS third_party/expat/lib/xmltok_ns.c:63:10
    #4 0x557820a3055c in prologProcessor third_party/expat/lib/xmlparse.c:4093:13
    #5 0x557820a30280 in prologInitProcessor third_party/expat/lib/xmlparse.c:3920:10
    #6 0x557820a2e0b5 in XML_ParseBuffer third_party/expat/lib/xmlparse.c:1893:25
    #7 0x557820a2ccb1 in XML_Parse third_party/expat/lib/xmlparse.c:1857:14
    #8 0x5578208e5755 in parse_entry third_party/stcorp_coda/libcoda/coda-definition-parse.c:4218:14
    #9 0x5578208e6957 in read_definition_file third_party/stcorp_coda/libcoda/coda-definition-parse.c:4263:9
    #10 0x5578208e6633 in coda_read_definitions third_party/stcorp_coda/libcoda/coda-definition-parse.c:4471:21
    #11 0x5578208a198d in LLVMFuzzerTestOneInput third_party/stcorp_coda/fuzz/coda_read_definitions_fuzzer.cc:19:3
    #12 0x5578214451b9 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) third_party/llvm/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:559:15
    #13 0x55782140b494 in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) third_party/llvm/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:301:6
    #14 0x557821419367 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) third_party/llvm/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:801:9
    #15 0x5578213ef5bd in main third_party/llvm/llvm-project/compiler-rt/lib/fuzzer/FuzzerMain.cpp:20:10
    #16 0x7f4b44823bbc in __libc_start_main (/usr/grte/v4/lib64/libc.so.6+0x38bbc)
    #17 0x5578208269e8 in _start /usr/grte/v4/debug-src/src/csu/../sysdeps/x86_64/start.S:108
SUMMARY: MemorySanitizer: use-of-uninitialized-value third_party/expat/lib/xmltok_impl.c:953:13 in normal_scanPoundName
Exiting
```

```
    parser_info_init(&info);
    info.zf = zf;
    info.entry_base_name = name;
    info.product_class = current_product_class;
    info.product_definition = current_product_definition;

    filesize = za_get_entry_size(entry);
    // The following allocates memory without initializing it. Using uninitialized memory can be potentially pose a security issue.
    info.buffer = malloc(filesize);
    if (info.buffer == NULL)
    {
        coda_set_error(CODA_ERROR_OUT_OF_MEMORY, "out of memory (could not allocate %lu bytes) (%s:%u)",
                       (long)filesize, __FILE__, __LINE__);
        parser_info_delete(&info);
        return -1;
    }
```

Test case: 
```
0000000: 5020613d0231223e 504b010237ffffff  P a=.1">PK..7...
0000010: ff8f080000f900fb 00cb00ff0a000000  ................
0000020: 010900002b00183c 214154544c494c54  ....+..<!ATTLILT
0000030: ff2885000000696e 6465782e786d6c00  .(....index.xml.
0000040: 00ff00000f291400 0003000000000000  .....)..........
0000050: 43ef2e3c21444e54 4954414e59ba00f1  C..<!DNTITANY...
0000060: 0c0c0c1f00000024 0001073f000100f6  .......$...?....
0000070: b20000504b1f0000 000000873f000070  ...PK.......?..p
0000080: 00004e02fb504b03 043334368f080000  ..N..PK..346....
0000090: f900fb00cb00ff0a 000000010900002b  ...............+
00000a0: 00080000f9252525 275b0000ff00002b  .....%%%'[.....+
00000b0: 3135373951e5e5f5 e52e27504b0506bf  1579Q.....'PK...
00000c0: 0000370100003300 0000160800000000  ..7...3.........
00000d0: fa0021d3d149532e 00000000000039e5  ..!..IS.......9.
00000e0: 504b0506bf000037 0100003300000016  PK.....7...3....
00000f0: 0800000000fa                       ......
```
[testcase-5633184541966336.zip](https://github.com/stcorp/coda/files/4950529/testcase-5633184541966336.zip)

**Comment by svniemeijer** [comment]
I can't replicate this problem. If I use your test case then I get:

```
codadd -D testcase-5633184541966336    
ERROR: could not read data from definition file: could not open testcase-5633184541966336 (not a zip file)

```

Which means it already bails out at `za_open` in `read_definition_file`. I don't understand how you get to `parse_entry`.

**Comment by ruimaranhao** [comment]
Apologies if this is a dumb question, but isn't this complaining about the fact that the file is not a zip one? Have you forgot to add .zip to the file name?

For the record, this is the fuzz driver:
```extern "C" {
int coda_read_definitions(const char *definition_path);
}
extern "C" int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
  FuzzerTemporaryFile temp_file(data, size);
  coda_init();
  auto done = absl::MakeCleanup([] { coda_done(); });
  coda_read_definitions(temp_file.filename());
  return 0;
}```

**Comment by svniemeijer** [comment]
I don't know how the fuzzer works, but CODA will check if the file starts with 'PK' (ziparchive.c:420) and otherwise bail out.
Your test file does not start with 'PK' (although there is a section 8 bytes in that starts with that).
A .zip extension is not needed (codadef files normally also have a .codadef extension)

**Comment by svniemeijer** [comment]
Actually, looking at it. Your test case is an xxd dump. Please provide the binary file.

**Comment by ruimaranhao** [comment]
Can you please try with the file attached. Thanks.

[testcase-5633184541966336.zip](https://github.com/stcorp/coda/files/4953593/testcase-5633184541966336.zip)

**Comment by svniemeijer** [comment]
Fixed in 22eb9197ab973055ce856f40b02157eb5833a958

Also improved check for proper zip file in 1d19ad61fd2011a54780335c896ceab4a7393cd3

**Comment by ruimaranhao** [comment]
Thanks, @svniemeijer !
